### PR TITLE
Milestone 6: Friendships v2 

### PR DESCRIPTION
**Pull request by JasemDuncan**
For this milestone is required to implement friendships with 2 rows for mutual friendship. 

In this case, we have a method named "confirm_friend(user)", in this method the we create the relation between friends.

The Project Specific in this part is:

As a logged-in user:
  
    I want to be able to accept or reject friendships invitation.

### CODE REVIEW

**Review by kevcap — CHANGES_REQUESTED** [review]
# [STATUS] CHANGES REQUIRED♻️

#### Hey team,  your project is looking really good, however, to accomplish as the actual project, you required to do these changes, listed below:
 
## [REQUIRED CHANGES]

- [x] You project has 33 examples, 9 failures, please fix the failures;
image

## Please make the required changes and resubmit for a new code review once you are ready💪

## As always, I wish you all the very best and see you both at the top🚀


👨🏻‍💻 **TSE: Kevin Alves**

  

📫 Have any question? Better call Kevin:

<a  href="mailto:kevinalves.js@gmail.com"> gmail</a> [Twitter Follow](https://twitter.com/intent/follow?screen_name=kevinnnmn) [](https://microverse-students.slack.com/team/UT86D3AP4)

<sub>*We're here to help you, to ensure that you're using the best practices and getting you ready for the job market. In real-world, you'll have a lot of changes requests for your real clients, it's good to get used to it.</sub>

**Comment by JasemDuncan** [comment]
Hello @kevincappuccino, I am writing on slack. I have some questions please.

**Comment by JasemDuncan** [comment]
The testing part should run it one file by one. that is because for the testing file we are using capybara, and the syntax will cause an error in the other testing files that are not using, that is why we have for some testing require 'rails_helper' and 'rails_helper_capybara'.
imagen

imagen

Please if there is a way to run all witout having the error about, please give us a hint. Thanks.

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi Team     
#### Great Job so far, everything works fine, but there are a few more things required. 
### General
- [ ] The test suite is still failing, let's fix it.


### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello @KingsleyMcSimon and @JasemDuncan 

You've both done an awesome job so far 👌. But there are still some changes that are necessary for this project to be approved.

In Friendship Version 2, we are going handle friendships a little more different

- [ ] When friendship is confirmed, an additional row with reversed user-friend pair is created
```ruby
# Friendship model

def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```

### - [ ] Thanks to this, we can find user friends only by user_id in the friendships table
```ruby
# User model

has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```

 - [ ] User#pending_friends can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)

```ruby
# User model

# Users who needs to confirm friendship
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
```

 - [ ] In order to get all posts written by a user on their friends, students can use a simple method in User model (in order to NOT do it with Enumerables)
```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends << self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```

#### kindly make the appropriate changes and ask for another code review 👍 

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

File: app/models/friendship.rb [hunk 1]
```diff
@@ -1,4 +1,10 @@
 class Friendship < ApplicationRecord
   belongs_to :user
   belongs_to :friend, class_name: 'User'
+  # def confirm_friend(user)
+  #   self.update_attributes(confirmed: true)
+  #   Friendship.create!(friend_id: self.user_id,
+  #                   user_id: self.friend_id,
+  #                   confirmed: true)
+  # end
```
Comment (ghost) [inline app/models/friendship.rb]: Remove the commented code.

File: app/models/post.rb [hunk 2]
```diff
@@ -8,11 +8,11 @@ class Post < ApplicationRecord
...
-    first_part = 'SELECT * FROM posts WHERE user_id IN'
-    Post.find_by_sql("#{first_part} (SELECT id FROM users WHERE id IN(#{us})) ORDER BY created_at")
-  end
+  # def timeline_posts(user)
+  #   us = user.friends.map(&:id)
+  #   us << user.id
+  #   us = us.join(',')
+  #   first_part = 'SELECT * FROM posts WHERE user_id IN'
+  #   Post.find_by_sql("#{first_part} (SELECT id FROM users WHERE id IN(#{us})) ORDER BY created_at")
+  # end
```
Comment (ghost) [inline app/models/post.rb]: Remove the commented code.

File: app/models/user.rb [hunk 3]
```diff
@@ -11,6 +11,13 @@ class User < ApplicationRecord
   has_many :likes, dependent: :destroy
   has_many :friendships
   has_many :inverse_friendships, class_name: 'Friendship', foreign_key: 'friend_id'
+  # has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
```
Comment (ghost) [inline app/models/user.rb]: Remove the commented code.

File: app/models/user.rb [hunk 4]
```diff
@@ -19,14 +26,14 @@ def friends
   end
 
   # Users who have yet to confirm friend request
-  def pending_friends
-    friendships.map { |friendship| friendship.friend unless friendship.confirmed }.compact
-  end
+  # def pending_friends
+  #   friendships.map { |friendship| friendship.friend unless friendship.confirmed }.compact
+  # end
```
Comment (ghost) [inline app/models/user.rb]: Remove the commented code. Please check for more commented codes especially in your `app/models` directory.

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required

## Hello team 😃,

**Great job implementing the requested changes 💪. However, there are still a few fixes needed.**

1). **Please add unit tests for your `helpers`**

2). **Check more in the comments below.**

[Optional]

- **You can add tests for your controllers.**

## Happy coding 🥂 

**PS: You can reach me on slack (Rose Wanjohi).**

**Review by tirthajyoti-ghosh — APPROVED** [review]
# Approved ✅

## You have done an incredible job making this project 👏🧨💪! Just perfect 💯! 

### You have met all the requirements of the project. Go ahead and merge it! Onto the next project 🚀 !

Happy coding 👨‍💻 

<hr>

**Tirthajyoti Ghosh**

[GitHub](https://github.com/tirthajyoti-ghosh) | [LinkedIn](https://www.linkedin.com/in/tirthajyoti-ghosh/) | [Twitter](https://twitter.com/terrific_ghosh)

**Pull request merged**